pkg/frontend/querymiddleware/spin_off_subqueries_queryable.go

```diff
@@ -110,62 +110,64 @@ func (q *spinOffSubqueriesQuerier) Select(ctx context.Context, _ bool, hints *st
 		if err != nil {
 			return storage.ErrSeriesSet(errors.Wrap(err, "failed to parse subquery step"))
 		}
+		var queryOffset time.Duration
+		if offsetStr == "" {
+			queryOffset = 0
+		} else if queryOffset, err = time.ParseDuration(offsetStr); err != nil {
+			return storage.ErrSeriesSet(errors.Wrap(err, "failed to parse subquery offset"))
+		}
 
-		end := q.req.GetEnd()
-		// Align query to absolute time by querying slightly into the future.
+		end := q.req.GetEnd() - queryOffset.Milliseconds()
+		// Subqueries are aligned on absolute time, while range queries are aligned on relative time.
+		// To match the same behavior, we can query slightly into the future.
+		// The extra data points aren't used because the subquery is aggregated with an _over_time-style function.
```

Comment:
Is this statement correct? Isn't the adjustment below small enough that it only introduces an additional point, and this is filtered out during evaluation because the PromQL engine (Prometheus' or MQE) will only include points from `start` to `start+queryRange`? 

Alternatively, we could use the same time range logic MQE uses, which avoids querying into the future: https://github.com/grafana/mimir/blob/266a393379b2c981a83557c5d66e56c97251ffeb/pkg/streamingpromql/query.go#L384-L398

Reply:
Nice! The MQE logic is better than mine. I replaced it